### PR TITLE
FoldScaleAxis became non-recursive

### PR DESCRIPTION
**Pull request by d-smirnov**
This PR migrates FoldScaleAxis optimization pass from ExprVisitor/ExprMutator to non-recursive MixedModeVisitor/MixedModeMutator. The specific transforming part itself is still recursive, however the underlying traversal machinery is now non-recursive.

### CODE REVIEW

**Comment by d-smirnov** [comment]
@mbrookhart

**Comment by d-smirnov** [comment]
bump. also cc @tqchen

**Comment by mbrookhart** [comment]
Sorry! Not sure how I lost this. I don't fully understand this pass, and it seems you've made some more complicated changes above moving it to mixed mode? I would appriciate @tqchen's review, since it was originally his code.

**Comment by tqchen** [comment]
Would be great if @vinx13 can also help to take a look

**Comment by d-smirnov** [comment]
>and it seems you've made some more complicated changes above moving it to mixed mode?
There is an extra minor change related to not doing the optimization if there were no messages produced by preparation sub-passes.

**Comment by d-smirnov** [comment]
bump

**Review by mbrookhart — COMMENTED** [review]
I'm happy with the mixed mode parts, but I'd still love to get @vinx13's eyes on it.

File: src/relay/transforms/fold_scale_axis.cc [hunk 1]
```diff
@@ -735,25 +733,41 @@ class BackwardTransformer : public ObjectRef {
   }
   using ContainerType = BackwardTransformerNode;
 };
-
-Expr BackwardTransformerNode::Transform(const CallNode* call_node, Message message, Expr scale) {
-  static const auto& ftransform = Op::GetAttrMap<FBackwardTransform>("FScaleAxisBackwardTransform");
-  auto f = ftransform.get(call_node->op, nullptr);
-  if (f != nullptr) {
+/*!
```
Comment (vinx13) [inline src/relay/transforms/fold_scale_axis.cc:736]: nit: add a blank line

File: src/relay/transforms/fold_scale_axis.cc [hunk 2]
```diff
@@ -735,25 +733,41 @@ class BackwardTransformer : public ObjectRef {
...
-    Expr new_expr = f(GetRef<Call>(call_node), message, scale, GetRef<BackwardTransformer>(this));
     memo_[call] = new_expr;
     return new_expr;
   } else {
     ICHECK(!message.defined()) << "outstanding scale";
-    return NormalCallTransform(call_node);
+    return this->Mutate(expr);
   }
 }
-
```
Comment (vinx13) [inline src/relay/transforms/fold_scale_axis.cc:756]: nit: keep the blank line here

**Review by vinx13 — APPROVED** [review]

File: src/relay/transforms/fold_scale_axis.cc (hunk 1 above)
Comment (d-smirnov) [reply src/relay/transforms/fold_scale_axis.cc:737]: Added

File: src/relay/transforms/fold_scale_axis.cc (hunk 2 above)
Comment (d-smirnov) [reply src/relay/transforms/fold_scale_axis.cc:756]: Added

**Comment by vinx13** [comment]
@d-smirnov There are some irrelevant errors on CI, Could you try again to get CI passed? Thanks

**Comment by mbrookhart** [comment]
Thanks @d-smirnov @vinx13 @tqchen

**Pull request merged**